### PR TITLE
Fixes #36337 - Update rake and clamp deps

### PR DESCRIPTION
**Pull request by ofedoren**
This PR updates `rake` to any latest available and updates `clamp` to be any version, but `2.0.0` (currently, the latest is 1.3.2).

There is also some refactoring due to `clamp` changes and some refactoring due to `rake` spamming a lot of Ruby warnings. I fixed majority of them (mostly it was `uninitialized variable` warning, although I'm not sure that's a good idea: https://bugs.ruby-lang.org/issues/17055#note-6)

Anyway, there is a new `warning` flag for `rake`'s `TestTask`, which suppresses all Ruby warnings. Without this flag quite a lot of our tests will fail since they depend on strict output. I left `RUBY_WARNINGS` env variable to be set to run the tests with Ruby warnings.

### CODE REVIEW

**Comment by ofedoren** [comment]
[test hammer]

File: lib/hammer_cli/logger.rb [hunk 1]
```diff
@@ -54,7 +54,7 @@ def self.initialize_logger(logger)
       log_dir = File.expand_path(HammerCLI::Settings.get(:log_dir) || DEFAULT_LOG_DIR)
       begin
         FileUtils.mkdir_p(log_dir, :mode => 0750)
-      rescue Errno::EACCES => e
+      rescue Errno::EACCES => _
```
Comment (ekohl) [inline lib/hammer_cli/logger.rb:57]: Repeated pattern, so only commenting once.
```suggestion
      rescue Errno::EACCES
```

**Comment by ofedoren** [comment]
Thanks, @ekohl, updated.

File: lib/hammer_cli/abstract.rb [hunk 2]
```diff
@@ -341,27 +341,27 @@ def exception_handler_class
     end
 
     def self.desc(desc=nil)
-      @desc = desc if desc
-      @desc
+      return @desc ||= nil unless desc
```
Comment (ekohl) [inline lib/hammer_cli/abstract.rb:344]: I think this is a lot harder to read. It's also a bit of an odd API where it's both a setter and a getter.

Can you separate the undefined variable changes to a separate commit? AFAIK they're warnings, not errors so it should be possible to update rake without addressing them immediately.

**Review by ekohl — CHANGES_REQUESTED** [review]
I think the undefined variables changes make the code harder to understand. It'd be good to separate those out and review them separately.

**Comment by ofedoren** [comment]
Thanks, @ekohl, I've splitted all the changes here to separate commits. I hope this will make review easier.

Also, since https://github.com/theforeman/hammer-cli-foreman/pull/613 contains quite similar changes, I'll update that one after this PR gets resolved/approved.

File: hammer_cli.gemspec [hunk 3]
```diff
@@ -27,7 +27,7 @@ EOF
   s.require_paths = ["lib"]
   s.executables = ['hammer', 'hammer-complete']
 
-  s.add_dependency 'clamp', '>= 1.1', '< 1.2.0'
+  s.add_dependency 'clamp', '>= 1.1', '< 2.0.0'
```
Comment (ekohl) [inline hammer_cli.gemspec:30]: You use the `line` method introduced in clamp 1.3.1 so that should be the lower bound. We'll need to deal with that in packaging, but I think that's ok.

File: lib/hammer_cli/utils.rb [hunk 4]
```diff
@@ -48,18 +48,6 @@ def wrap(line_width: 80, break_sequence: "\n")
   end
 end
 
-class Hash
-  # for ruby < 2.5.0
```
Comment (ekohl) [inline lib/hammer_cli/utils.rb:52]: Does it make sense to set `required_ruby_version` to `>= 2.5` in the gemspec?

File: lib/hammer_cli/abstract.rb [hunk 5]
```diff
@@ -341,27 +341,27 @@ def exception_handler_class
...
     end
 
     def self.command_names(*names)
-      @names = names unless names.empty?
-      @names || (superclass.respond_to?(:command_names) ? superclass.command_names : nil)
+      names = names.compact
+      return @names ||= (superclass.respond_to?(:command_names) ? superclass.command_names : nil) if names.empty?
+
+      @names ||= []
+      @names |= names
```
Comment (ekohl) [inline lib/hammer_cli/abstract.rb:358]: This changes the implementation and now caches the response of the superclass. Is that intentional? I'd keep the change minimal:
```suggestion
      @names ||= nil
      @names = names unless names.empty?
      @names || (superclass.respond_to?(:command_names) ? superclass.command_names : nil)

**Review by ekohl — COMMENTED** [review]
https://stackoverflow.com/questions/2098742/ruby-instance-variable-not-initialized-warning/2098835#2098835 suggests the undefined warning goes away with Ruby 3.0. Since you disable warnings anyway, perhaps the last commit should be trimmed down or even skipped.

File: lib/hammer_cli/utils.rb (hunk 4 above)
Comment (ofedoren) [reply lib/hammer_cli/utils.rb:52]: Well, I'd set it to `>= 2.7` even, since I'm not sure we even test lower versions...

**Comment by ofedoren** [comment]
> suggests the undefined warning goes away with Ruby 3.0

And I found the commit with that :) https://github.com/ruby/ruby/commit/01b7d5acc702df22d306ae95f1a9c3096e63e624

So, I've decided to drop the last commit. Hammer is already slow, it's not worth to slow it more.

Updated, but without new changes related to uninitialized variables.

File: lib/hammer_cli/subcommand.rb [hunk 6]
```diff
@@ -11,6 +11,7 @@ def initialize(names, description, subcommand_class, options = {})
         @subcommand_class = subcommand_class
         @hidden = options[:hidden]
         @warning = options[:warning]
+        super(@names, @description, @subcommand_class)
```
Comment (ekohl) [inline lib/hammer_cli/subcommand.rb:14]: If you call this, I wonder if names, description and subclass_command aren't already stored. I think you can drop `@names =` etc.

**Review by ekohl — APPROVED** [review]

File: lib/hammer_cli/subcommand.rb (hunk 6 above)
Comment (ofedoren) [reply lib/hammer_cli/subcommand.rb:14]: Not really, AFAIR I've added this line because something wasn't initialized in clamp itself, crashing the commands.

Also, I rather call `super` with explicit arguments because implicitly it would send `params` as well, which is not supported by clamp.
